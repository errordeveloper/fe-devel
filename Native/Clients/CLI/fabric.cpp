/*
 *  Copyright 2010-2012 Fabric Engine Inc. All rights reserved.
 */

#include <Fabric/Clients/CLI/ClientFactory.h>
#include <Fabric/Base/Util/Log.h>
#include <Fabric/Core/Build.h>

#include <time.h>

namespace Fabric
{
  namespace CLI
  {
    extern "C" FABRIC_CLI_EXPORT void init( v8::Handle<v8::Object> target )
    {
      FABRIC_LOG( "%s version %s", Fabric::buildName, Fabric::buildFullVersion );

      v8::HandleScope handleScope;
      target->Set( v8::String::New("createClient"), CreateClientV8Function() );
    }
  }
}
