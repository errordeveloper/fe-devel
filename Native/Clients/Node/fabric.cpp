/*
 *  Copyright 2010-2012 Fabric Engine Inc. All rights reserved.
 */

#include <Fabric/Clients/Node/ClientFactory.h>
#include <Fabric/Base/Util/Log.h>
#include <Fabric/Core/Build.h>

#include <llvm/Support/Threading.h>
#include <time.h>
#include <stdlib.h>

namespace Fabric
{
  namespace CLI
  {
    extern "C" FABRIC_CLI_EXPORT void init( v8::Handle<v8::Object> target )
    {
      FABRIC_LOG( "%s version %s", Fabric::buildName, Fabric::buildFullVersion );

      llvm::llvm_start_multithreaded();
      if ( !llvm::llvm_is_multithreaded() )
      {
        FABRIC_LOG( "LLVM not compiled with multithreading enabled; aborting" );
        abort();
      }

      v8::HandleScope handleScope;
      target->Set( v8::String::New("createClient"), CreateClientV8Function() );
    }
  }
}
