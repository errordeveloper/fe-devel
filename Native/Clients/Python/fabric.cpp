/*
 *  Copyright 2010-2012 Fabric Engine Inc. All rights reserved.
 */

#include <Fabric/Base/Util/SimpleString.h>
#include <Fabric/Base/Exception.h>
#include <Fabric/Clients/Python/ClientWrap.h>
#include <Fabric/Base/Util/Log.h>
#include <Fabric/Core/Build.h>

#include <map>
#include <time.h>
#include <llvm/Support/Threading.h>

//#define FABRIC_PYTHON_DEBUG

namespace Fabric
{
  namespace Python
  {
    extern "C" FABRIC_CLI_EXPORT void initialize()
    {
      FABRIC_LOG( "%s version %s", Fabric::buildName, Fabric::buildFullVersion );

      llvm::llvm_start_multithreaded();
      if ( !llvm::llvm_is_multithreaded() )
      {
        FABRIC_LOG( "LLVM not compiled with multithreading enabled; aborting" );
        abort();
      }
    }

    extern "C" FABRIC_CLI_EXPORT void createClient( void **ptr, char const *opts )
    {
      ClientWrap *client = new ClientWrap( opts );

#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling createClient: %lx", client);
#endif

      client->retain();
      *ptr = client;
    }
    
    extern "C" FABRIC_CLI_EXPORT void jsonExec( void *client_, char const *data, size_t length, const char **result )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling jsonExec: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );

#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "sending to jsonExec: %s", data );
#endif

      client->jsonExecAndAllocCStr( data, length, result );

#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "received from jsonExec: %s", *result );
#endif
    }

    extern "C" FABRIC_CLI_EXPORT void freeString( void *client_, void *string )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling freeString: %lx (%lx)", client_, string );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->freeJsonCStr( (const char *)string );
    }

    extern "C" FABRIC_CLI_EXPORT void setJSONNotifyCallback(
        void *client_,
        void (*callback)(const char *)
        )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling setJSONNotifyCallback: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->setJSONNotifyCallback( callback );
    }

    extern "C" FABRIC_CLI_EXPORT void freeClient( void *client_ )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling close: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->release();
    }

    extern "C" FABRIC_CLI_EXPORT void runScheduledCallbacks( void *client_ )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling runScheduledCallbacks: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->runScheduledCallbacks();
    }
  };
};
