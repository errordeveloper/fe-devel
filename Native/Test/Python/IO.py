import fabric
fabricClient = fabric.createClient()

try:
  fabricClient.IO.putTextFileContent("dummy", "badHandle")
except Exception as e:
  print('Bad handle error: ' + str(e))

readOnlyFileAndFolderHandles = fabricClient.IO.queryUserFileAndFolderHandle(fabricClient.IO.forOpen, "dummy title", "txt", "testfile")

try: 
  fabricClient.IO.putTextFileContent(readOnlyFileAndFolderHandles['file'], "dummy")
except Exception as e:
  print('No write access error: ' + str(e))


fileAndFolderHandles = fabricClient.IO.queryUserFileAndFolderHandle(fabricClient.IO.forOpenWithWriteAccess, "dummy title", "txt", "testfile")
print('User file name: ' + fabricClient.IO.getFileHandleInfo(fileAndFolderHandles['file'])['fileName'])

relativeFileHandle = fabricClient.IO.buildFileHandleFromRelativePath(fileAndFolderHandles['folder'] + '/SubDir/testfile2.txt')

try:
  dummy = fabricClient.IO.getTextFileContent(relativeFileHandle)
except Exception as e:
  print('File not found error')#'e' has random handle name; don't display it

fabricClient.IO.putTextFileContent(relativeFileHandle,"FileBody")
body = fabricClient.IO.getTextFileContent(relativeFileHandle)
print('File content: ' + body)

fabricClient.IO.putTextFileContent(relativeFileHandle," with appended text",True)
body = fabricClient.IO.getTextFileContent(relativeFileHandle)
print('File content with appended text: ' + body)

fileInfo = fabricClient.IO.getFileHandleInfo(relativeFileHandle)
print('File info: ' + fabric.stringify(fileInfo))

node = fabricClient.DependencyGraph.createResourceLoadNode("node")

myvars = {}
myvars['callbackStep'] = 0

def loadFailureCallback( selfNode ):
  if myvars['callbackStep'] == 0:
    print( 'bad url load failure' )
    node.setData('url',0,relativeFileHandle)
    myvars['callbackStep'] = 1
    node.evaluate()
  else:
    print('failure callback: Unexpected')
    fabricClient.close()

node.addOnLoadFailureCallback( loadFailureCallback )

def loadSuccessCallback( selfNode ):
  print( 'url file load succeeded' )
  if myvars['callbackStep'] == 1:
    resourceResultFileHandle = fabricClient.IO.buildFileHandleFromRelativePath(fileAndFolderHandles['folder'] + '/SubDir/testfile3.txt')
    node.putResourceToFile(resourceResultFileHandle,'resource')
    body = fabricClient.IO.getTextFileContent(resourceResultFileHandle)
    print('Resource data saved to file: ' + body)

    sourceFileHandle = fabricClient.IO.buildFileHandleFromRelativePath(fileAndFolderHandles['folder'] + '/SubDir/testfile4.txt')
    fabricClient.IO.putTextFileContent(sourceFileHandle,"Resource source data from direct file")
    fabricClient.flush()

    #Load from explicit file location, to a file handle
    myvars['callbackStep'] = 2
    node.setData('storeDataAsFile',0,True)
    fabricClient.flush()
    node.setData('url',0,'testfile://TMP/SubDir/testfile4.txt')
    fabricClient.flush()
    node.evaluate()
  elif myvars['callbackStep'] == 2:
    resourceResult = node.getData('resource',0)
    resourceFileContent = fabricClient.IO.getTextFileContent(resourceResult['dataExternalLocation'])
    print('Resource file content: ' + resourceFileContent)

    fabricClient.close()
  else:
    print('success callback: Unexpected')
    fabricClient.close()

node.addOnLoadSuccessCallback( loadSuccessCallback )

node.setData('url',0,'badURL')
node.evaluate()

