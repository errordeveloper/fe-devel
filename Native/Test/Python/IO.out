Bad handle error: Fabric core exception: IO.putTextFileContent('{"content": "badHandle", "append...'): Error: ambiguous: handle's relative path could be a file or a folder; first create a file or folder handle from relative path
No write access error: Fabric core exception: IO.putTextFileContent('{"content": "dummy", "append": f...'): File has no write access
File and folder handles: [{"file": "file://TMP/default.txt"}, {"folder": "file://TMP"}]
User file name: default.txt
File not found error
File content: FileBody
File content with appended text: FileBody with appended text
File info: [{"exists": true}, {"fileName": "testfile2.txt"}, {"fileSize": 27}, {"type": "file"}, {"writeAccess": true}]
bad url load failure
url file load succeeded
Resource data saved to file: FileBody with appended text
url file load succeeded
Resource file content: Resource source data from direct file
